From c9b734e9613ed2bf077b84a4927ed88c3c40d23c Mon Sep 17 00:00:00 2001
From: Paul Naoumenko <paulx.naoumenko@intel.com>
Date: Thu, 17 Oct 2013 17:49:35 +0200
Subject: [PORT FROM MAIN]wifi: Differentiate bcm chips firmware path

BZ: 83712

In the case of broadcom wireless chip, we need to know which chip is used
in order to send back the proper firmware path. This is necessary to add
support for dual chip.

Change-Id: I95619a3cafe7a31c9b7f5402f14d5fa6556f78ea
Orig-Change-Id: Icbab3813ec417b55fcf7572bb7b68570401081cd
Category: device enablement
Domain: CWS.WIFI-Bluewoods
Origin: upstream
Upstream-Candidate: no,proprietary
Signed-off-by: Quentin Casasnovas <quentinx.casasnovas@intel.com>
---
 wifi/Android.mk |   13 +++++++++++++
 wifi/wifi.c     |   44 +++++++++++++++++++++++++++++++++++++++++---
 2 files changed, 54 insertions(+), 3 deletions(-)

diff --git a/wifi/Android.mk b/wifi/Android.mk
index daed62d..bf33e1d 100644
--- a/wifi/Android.mk
+++ b/wifi/Android.mk
@@ -16,15 +16,28 @@ endif
 ifdef WIFI_FIRMWARE_LOADER
 LOCAL_CFLAGS += -DWIFI_FIRMWARE_LOADER=\"$(WIFI_FIRMWARE_LOADER)\"
 endif
+
 ifdef WIFI_DRIVER_FW_PATH_STA
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_STA=\"$(WIFI_DRIVER_FW_PATH_STA)\"
 endif
+LOCAL_CFLAGS += -DWIFI_DRIVER_43241_FW_PATH_STA=\"$(WIFI_DRIVER_43241_FW_PATH_STA)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4334_FW_PATH_STA=\"$(WIFI_DRIVER_4334_FW_PATH_STA)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4335_FW_PATH_STA=\"$(WIFI_DRIVER_4335_FW_PATH_STA)\"
+
 ifdef WIFI_DRIVER_FW_PATH_AP
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_AP=\"$(WIFI_DRIVER_FW_PATH_AP)\"
 endif
+LOCAL_CFLAGS += -DWIFI_DRIVER_43241_FW_PATH_AP=\"$(WIFI_DRIVER_43241_FW_PATH_AP)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4334_FW_PATH_AP=\"$(WIFI_DRIVER_4334_FW_PATH_AP)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4335_FW_PATH_AP=\"$(WIFI_DRIVER_4335_FW_PATH_AP)\"
+
 ifdef WIFI_DRIVER_FW_PATH_P2P
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_P2P=\"$(WIFI_DRIVER_FW_PATH_P2P)\"
 endif
+LOCAL_CFLAGS += -DWIFI_DRIVER_43241_FW_PATH_P2P=\"$(WIFI_DRIVER_43241_FW_PATH_P2P)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4334_FW_PATH_P2P=\"$(WIFI_DRIVER_4334_FW_PATH_P2P)\"
+LOCAL_CFLAGS += -DWIFI_DRIVER_4335_FW_PATH_P2P=\"$(WIFI_DRIVER_4335_FW_PATH_P2P)\"
+
 ifdef WIFI_DRIVER_FW_PATH_PARAM
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_PARAM=\"$(WIFI_DRIVER_FW_PATH_PARAM)\"
 endif
diff --git a/wifi/wifi.c b/wifi/wifi.c
index bebdb5e..77995e9 100644
--- a/wifi/wifi.c
+++ b/wifi/wifi.c
@@ -109,6 +109,12 @@ static const char IFNAME[]              = "IFNAME=";
 #define IFNAMELEN			(sizeof(IFNAME) - 1)
 static const char WPA_EVENT_IGNORE[]    = "CTRL-EVENT-IGNORE ";
 
+/*
+ * This gets defined by the script load_bcmdriver in
+ * vendor/intel/common/wifi/bcm_specific/
+ */
+static const char BCM_PROP_CHIP[]	= "wlan.bcm.chip";
+
 static const char SUPP_ENTROPY_FILE[]   = WIFI_ENTROPY_FILE;
 static unsigned char dummy_key[21] = { 0x02, 0x11, 0xbe, 0x33, 0x43, 0x35,
                                        0x68, 0x47, 0x84, 0x99, 0xa9, 0x2b,
@@ -864,14 +870,46 @@ int wifi_command(const char *command, char *reply, size_t *reply_len)
 
 const char *wifi_get_fw_path(int fw_type)
 {
+    char bcm_prop_chip[PROPERTY_VALUE_MAX];
+
     switch (fw_type) {
     case WIFI_GET_FW_PATH_STA:
-        return WIFI_DRIVER_FW_PATH_STA;
+	if (property_get(BCM_PROP_CHIP, bcm_prop_chip, NULL)) {
+	    if (strstr(bcm_prop_chip, "43241"))
+		return WIFI_DRIVER_43241_FW_PATH_STA;
+	    else if (strstr(bcm_prop_chip, "4334"))
+		return WIFI_DRIVER_4334_FW_PATH_STA;
+	    else if (strstr(bcm_prop_chip, "4335"))
+		return WIFI_DRIVER_4335_FW_PATH_STA;
+	}
+	else
+	    return WIFI_DRIVER_FW_PATH_STA;
     case WIFI_GET_FW_PATH_AP:
-        return WIFI_DRIVER_FW_PATH_AP;
+	if (property_get(BCM_PROP_CHIP, bcm_prop_chip, NULL)) {
+	    if (strstr(bcm_prop_chip, "43241"))
+		return WIFI_DRIVER_43241_FW_PATH_AP;
+	    else if (strstr(bcm_prop_chip, "4334"))
+		return WIFI_DRIVER_4334_FW_PATH_AP;
+	    else if (strstr(bcm_prop_chip, "4335"))
+		return WIFI_DRIVER_4335_FW_PATH_AP;
+	}
+	else
+	    return WIFI_DRIVER_FW_PATH_AP;
     case WIFI_GET_FW_PATH_P2P:
-        return WIFI_DRIVER_FW_PATH_P2P;
+	if (property_get(BCM_PROP_CHIP, bcm_prop_chip, NULL)) {
+	    if (strstr(bcm_prop_chip, "43241"))
+		return WIFI_DRIVER_43241_FW_PATH_P2P;
+	    else if (strstr(bcm_prop_chip, "4334"))
+		return WIFI_DRIVER_4334_FW_PATH_P2P;
+	    else if (strstr(bcm_prop_chip, "4335"))
+		return WIFI_DRIVER_4335_FW_PATH_P2P;
+	}
+	else
+	    return WIFI_DRIVER_FW_PATH_P2P;
+    default:
+	    ALOGE("Unknown firmware type (%d)", fw_type);
     }
+
     return NULL;
 }
 
-- 
1.7.4.1

